invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -0,0 +1,341 @@
+        layer-specific are passed to __init__().
+        Args:
+            ip_adapter_weights: The IP-Adapter attention weights. ip_adapter_weights[i] contains the attention weights
+                for the i'th IP-Adapter.
+        """
+
+        self._ip_adapter_attention_weights = ip_adapter_attention_weights
+        self.attention_type, self.slice_size = self._select_attention()
+
+    def _select_attention(self):
+        config = get_config()
+        attention_type = config.attention_type
+        if attention_type in ["normal", "xformers"]:
+            logger.warning(f'Attention "{attention_type}" no longer supported, "torch-sdp" will be used instead.')
+            attention_type = "torch-sdp"
+
+        if attention_type == "auto":
+            exec_device = TorchDevice.choose_torch_device()
+            if exec_device.type == "mps":
+                attention_type = "sliced"
+            else:
+                attention_type = "torch-sdp"
+
+        if attention_type == "torch-sdp" and not hasattr(F, "scaled_dot_product_attention"):
+            raise ImportError("torch-sdp attention requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
+
+        slice_size = None
+        if attention_type == "sliced":
+            slice_size = config.attention_slice_size
+            if slice_size not in ["auto", "balanced", "max"] and not isinstance(slice_size, int):
+                raise ValueError(f"Unsupported attention_slice_size: {slice_size}")
+            if slice_size == "balanced":
+                slice_size = "auto"
+
+        return attention_type, slice_size
+
+    def __call__(
+        self,
+        attn: Attention,
+        hidden_states: torch.Tensor,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        temb: Optional[torch.Tensor] = None,
+        # For Regional Prompting:
+        regional_prompt_data: Optional[RegionalPromptData] = None,
+        percent_through: Optional[torch.Tensor] = None,
+        # For IP-Adapter:
+        regional_ip_data: Optional[RegionalIPData] = None,
+        *args,
+        **kwargs,
+    ) -> torch.Tensor:
+        # If true, we are doing cross-attention, if false we are doing self-attention.
+        is_cross_attention = encoder_hidden_states is not None
+
+        residual = hidden_states
+
+        if attn.spatial_norm is not None:
+            hidden_states = attn.spatial_norm(hidden_states, temb)
+
+        input_ndim = hidden_states.ndim
+
+        if input_ndim == 4:
+            batch_size, channel, height, width = hidden_states.shape
+            hidden_states = hidden_states.view(batch_size, channel, height * width).transpose(1, 2)
+
+        batch_size, key_length, _ = (
+            hidden_states.shape if encoder_hidden_states is None else encoder_hidden_states.shape
+        )
+        query_length = hidden_states.shape[1]
+
+        attention_mask = self.prepare_attention_mask(
```

Comment:
Because the prep function now needs a mandatory attention mask, the conditional needs to be added here. This is called only when `if attention_mask is not None`